vertx-web-session-stores/vertx-web-sstore-caffeine/src/main/java/io/vertx/ext/web/sstore/caffeine/impl/CaffeineSessionStoreImpl.java

```diff
@@ -0,0 +1,140 @@
+package io.vertx.ext.web.sstore.caffeine.impl;
+
+import com.github.benmanes.caffeine.cache.AsyncCache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.Expiry;
+import io.vertx.codegen.annotations.Nullable;
+import io.vertx.core.Future;
+import io.vertx.core.Vertx;
+import io.vertx.core.internal.ContextInternal;
+import io.vertx.core.internal.VertxInternal;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.auth.prng.VertxContextPRNG;
+import io.vertx.ext.web.Session;
+import io.vertx.ext.web.sstore.AbstractSession;
+import io.vertx.ext.web.sstore.SessionStore;
+import io.vertx.ext.web.sstore.caffeine.CaffeineSessionStore;
+import io.vertx.ext.web.sstore.impl.SharedDataSessionImpl;
+
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * @author <a href="mailto:lazarbulic@gmail.com">Lazar Bulic</a>
+ */
+public class CaffeineSessionStoreImpl implements SessionStore, CaffeineSessionStore {
+
+  /**
+   * Default name for map used to store sessions
+   */
+  private static final String DEFAULT_SESSION_MAP_NAME = "vertx-web.caffeine.sessions";
+
+
+  private AsyncCache<String, Session> localCaffeineCache;
+  private VertxContextPRNG random;
+  private String sessionMapName;
+
+  private VertxInternal vertx;
+
+  public CaffeineSessionStoreImpl() {
+    // required for the service loader
+  }
+
+  @Override
+  public Session createSession(long timeout) {
+    return new SharedDataSessionImpl(random, timeout, DEFAULT_SESSIONID_LENGTH);
+  }
+
+  @Override
+  public Session createSession(long timeout, int length) {
+    return new SharedDataSessionImpl(random, timeout, length);
+  }
+
+  @Override
+  public SessionStore init(Vertx vertx, JsonObject options) {
+    // initialize a secure random
+    this.random = VertxContextPRNG.current(vertx);
+    this.vertx = (VertxInternal) vertx;
+    this.sessionMapName = options.getString("mapName", DEFAULT_SESSION_MAP_NAME);
+    final ContextInternal ctx = ((VertxInternal) vertx).getOrCreateContext();
+    localCaffeineCache = (AsyncCache<String, Session>) ctx.contextData()
+      .computeIfAbsent(sessionMapName, k -> Caffeine.newBuilder()
+        .executor(cmd -> ctx.runOnContext(v -> cmd.run()))
+        .expireAfter(Expiry.creating((String key, Session session) ->
+          Duration.ofMillis(session.timeout())))
+        .buildAsync());
+    return this;
```

Comment:
I don't understand why we can't create an async cache directly, can you explain?

Reply:
I wanted to store the cache somewhere where it can be accessed from multiple parts of code and reused. This is to achieve the same thing as `vertx.sharedData().getLocalMap` that i used for LocalSessionStoreImpl

Reply:
Shared data can be accessed from any thread provided you have a Vert.x instamce, context data is only available if you have a reference to the context instance. So I don't believe storing the cache in context data will work, for example if several instances of the verticle are deployed.

Regarding access to the session cache or map from external map, have you got an example use case?

Reply:
So if I understand correctly if I have 2 vertices deployed ((VertxInternal) vertx).getOrCreateContext(); will not result in same context and hence have different context data? In this case current implementation would have different instances of caffeine cache?  

Use case where this cache needs to be shared is for example deploying multiple instances of same Verticle. If cache is not shared between them then session is rendered useless. This is resolved in https://github.com/vert-x3/vertx-web/blob/10fcd1f668dc1aceb0f5d9fb0e4ab4b030b08fb9/vertx-web/src/main/java/io/vertx/ext/web/sstore/impl/LocalSessionStoreImpl.java#L77

Issues is that Caffeine caches can not be put in shareData

Reply:
Yes, contexts (thus related data) would be different in different instances of a verticle.

Right, that use case you describe is actually what's needed for the `LocalSessionStore` implementation. But other than that, I'm not sure why we'd need to access the shared local map from the outside.

Reply:
There is no need to access it outside of the LocalSessionImpl, there is just need to share it across different Vertices.

Reply:
Ok, then I think you could use `io.vertx.core.impl.VertxImpl#createSharedResource`